_posts/2022-09-08-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,168 @@
+
+However, when using via-IR code generation the entire contract is first translated to Yul and then optimized as a whole. In this case, 1-3 can happen in Solidity code and only the function called in (2) needs to involve an inline assembly block calling ``return(...)`` or ``stop()`` (the compiler will not generate ``return(...)`` or ``stop()`` instructions that can trigger the bug without the use of inline assembly). Note that due to inlining, this inline assembly block can also merely occur in another nested function call, i.e. any call in (2) that, through any chain of nested calls, can both return to the caller and terminate via assembly is susceptible.
+
+The following is a minimal example that will exhibit the bug if compiled via IR with enabled optimizer:
+
+```solidity
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1; // This write used to be incorrectly removed by the Yul optimizer due to the bug.  
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		// The relevant part of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which the transaction is terminated (i.e. if a is false).
+		// In this case the write x = 1 above will be missing.
+		if (a) return;
+		assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via IR with enabled optimizer will result in ``f(false)`` to incorrectly terminate the transaction without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic optimizer step, e.g. the following contract is almost identical, but unaffected, since ``g()`` will be inlined (in the snippet above the Solidity-level ``return;`` prevents easy inlining):
+
+```solidity
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1;
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		if (!a)
+			assembly { return(0,0) }
+	}
+}
+```
+
+To check whether your contract is actually affected, you need to trigger the condition leading to the assembly ``return(...)`` or ``stop()`` and verify that all storage writes that should have happened in the code leading to it were correctly performed.
+
+## Technical Details
+
+The Yul optimizer step responsible for the bug is the Unused Store Eliminator (abbreviated as `S` in the optimizer step sequence). It is meant to remove storage writes that it can determine to be redundant. A storage write is considered redundant in either of the following two cases:
```

Comment:
```suggestion
The Yul optimizer step responsible for the bug is the Unused Store Eliminator (abbreviated as `S` in the [optimizer step sequence](https://docs.soliditylang.org/en/latest/yul.html#optimization-step-sequence)). It is meant to remove storage writes that it can determine to be redundant. A storage write is considered redundant in either of the following two cases:
```